### PR TITLE
feat: add MemorystoreChatMessageHistory

### LINKED ISSUE #7: Implementation of chat message history for GCP Memorystore for Redis
Langchain integration of GCP Memorystore for Redis

### PR DESCRIPTION
**Pull request by craig08**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/langchain-google-memorystore-redis-python/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #7 🦕

### CODE REVIEW

File: src/langchain_google_memorystore_redis/__init__.py [hunk 1]
```diff
@@ -11,3 +11,7 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
+from langchain_google_memorystore_redis.redis_chat_message_history import MemorystoreChatMessageHistory
```
Comment (PingXie) [inline src/langchain_google_memorystore_redis/__init__.py:15]: should we name the file "chat_message_history.py". We already have redis in the package name to differentiate the engine. I would imagine when we started integrating with a different engine say memcached (if ever :-)), we would create a new package name like langchain_google_memorystore_memcached. What do you think?

File: src/langchain_google_memorystore_redis/redis_chat_message_history.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    BaseMessage,
+    message_to_dict,
+    messages_from_dict,
+)
+
+
+class MemorystoreChatMessageHistory(BaseChatMessageHistory):
+    """Chat message history stored in a Cloud Memorystore for Redis database."""
+
+    def __init__(
```
Comment (PingXie) [inline src/langchain_google_memorystore_redis/redis_chat_message_history.py:30]: I think the recommendation is to load all messages in the constructor.

Comment (craig08) [reply src/langchain_google_memorystore_redis/redis_chat_message_history.py]: SG renamed the file

Comment (craig08) [reply src/langchain_google_memorystore_redis/redis_chat_message_history.py]: Synced offline. Added a list to cache the history.

File: src/langchain_google_memorystore_redis/chat_message_history.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                expire and be eliminated from the Redis instance since the most
+                recent message is added.
+        """
+
+        self.redis = client
+        self.key = session_id
+        self.ttl = ttl
+
+        all_elements = self.redis.lrange(self.key, 0, -1)
+        self.history = messages_from_dict(
```
Comment (averikitsch) [inline src/langchain_google_memorystore_redis/chat_message_history.py:51]: The message property below should manage the loading of the messages, therefore the history property should not be used.

**Review by averikitsch — DISMISSED** [review]

File: src/langchain_google_memorystore_redis/chat_message_history.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            ttl: Specifies the time in seconds after which the session will
+                expire and be eliminated from the Redis instance since the most
+                recent message is added.
+        """
+
+        self.redis = client
+        self.key = session_id
+        self.ttl = ttl
+
+        all_elements = self.redis.lrange(self.key, 0, -1)
```
Comment (PingXie) [inline src/langchain_google_memorystore_redis/chat_message_history.py:50]: should we try/except Redis errors?

File: src/langchain_google_memorystore_redis/chat_message_history.py (hunk 3 above)
Comment (craig08) [reply src/langchain_google_memorystore_redis/chat_message_history.py:51]: The history property is just a cache of the all message history. Do you mean we can always query the Redis server for every `.messages` command? This is what I originally implemented, can you check whether commit [b033836](https://github.com/googleapis/langchain-google-memorystore-redis-python/pull/10/commits/b0338360ad1ee11017624b4674bed0db6482143a) does what you meant?

File: src/langchain_google_memorystore_redis/chat_message_history.py (hunk 4 above)
Comment (craig08) [reply src/langchain_google_memorystore_redis/chat_message_history.py:50]: Redis exception is directly thrown out to the client application if it encounters any issue. This aligns with the community practice like: https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/chat_message_histories/elasticsearch.py#L106-L116. They directly throw out exceptions if any setup is invalid.

In our example, if the user provides an invalid address, they will see direct exception from redis-py:

```
>>> c2 = MemorystoreChatMessageHistory(redis.from_url('redis://localhost:6378'), 'chat1')
Traceback (most recent call last):
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/connection.py", line 264, in connect
    sock = self.retry.call_with_retry(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/retry.py", line 46, in call_with_retry
    return do()
           ^^^^
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/connection.py", line 265, in <lambda>
    lambda: self._connect(), lambda error: self.disconnect(error)
            ^^^^^^^^^^^^^^^
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/connection.py", line 627, in _connect
    raise err
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/connection.py", line 615, in _connect
    sock.connect(socket_address)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/google/home/craigchi/venv/langchain/langchain-google-memorystore-redis-python/src/langchain_google_memorystore_redis/chat_message_history.py", line 50, in __init__
    all_elements = self.redis.lrange(self.key, 0, -1)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/commands/core.py", line 2754, in lrange
    return self.execute_command("LRANGE", name, start, end)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/client.py", line 533, in execute_command
    conn = self.connection or pool.get_connection(command_name, **options)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/connection.py", line 1086, in get_connection
    connection.connect()
  File "/usr/local/google/home/craigchi/venv/langchain/lib/python3.11/site-packages/redis/connection.py", line 270, in connect
    raise ConnectionError(self._error_message(e))
redis.exceptions.ConnectionError: Error 111 connecting to localhost:6378. Connection refused.
```

File: src/langchain_google_memorystore_redis/chat_message_history.py (hunk 3 above)
Comment (craig08) [reply src/langchain_google_memorystore_redis/chat_message_history.py]: Synced offline. Reverted the commits that adds the history property

**Review by PingXie — APPROVED** [review]

File: src/langchain_google_memorystore_redis/chat_message_history.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            client: A redis.Redis client object.
+            session_id: A string that uniquely identifies the chat history.
+            ttl: Specifies the time in seconds after which the session will
+                expire and be eliminated from the Redis instance since the most
+                recent message is added.
+        """
+
+        self.redis = client
+        self.key = session_id
+        self.ttl = ttl
```
Comment (kurtisvg) [inline src/langchain_google_memorystore_redis/chat_message_history.py:48]: nit: any reason for these to not be internal? 
```suggestion
        self._redis = client
        self._key = session_id
        self._ttl = ttl
```

File: src/langchain_google_memorystore_redis/chat_message_history.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    def __del__(self):
+        self.redis.close()
```
Comment (kurtisvg) [inline src/langchain_google_memorystore_redis/chat_message_history.py:51]: If a user passes this in, should we really close it automatically?

File: src/langchain_google_memorystore_redis/chat_message_history.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+
+class MemorystoreChatMessageHistory(BaseChatMessageHistory):
+    """Chat message history stored in a Cloud Memorystore for Redis database."""
+
+    def __init__(
+        self,
+        client: redis.Redis,
+        session_id: str,
+        ttl: Optional[int] = None,
```
Comment (kurtisvg) [inline src/langchain_google_memorystore_redis/chat_message_history.py:34]: Any reason to not spell out "time_to_live"? 

Is it worth using the same terminology redis does -- "expire"?

File: src/langchain_google_memorystore_redis/chat_message_history.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+    def __del__(self):
+        self.redis.close()
+
+    @property
+    def messages(self) -> List[BaseMessage]:
+        """Retrieve all messages chronologically stored in this session."""
+        all_elements = self.redis.lrange(self.key, 0, -1)
+        messages = messages_from_dict(
+            [json.loads(e.decode("utf-8")) for e in all_elements]
```
Comment (kurtisvg) [inline src/langchain_google_memorystore_redis/chat_message_history.py:58]: any reason to pass in "utf-8" specifically?

**Review by kurtisvg — CHANGES_REQUESTED** [review]

File: src/langchain_google_memorystore_redis/chat_message_history.py (hunk 7 above)
Comment (craig08) [reply src/langchain_google_memorystore_redis/chat_message_history.py:34]: I'm okay to different terms but I generally think TTL is more commonly used in the context of Redis - https://redis.io/commands/ttl/ which is a command to understand how many seconds a key can live in Redis.

File: src/langchain_google_memorystore_redis/chat_message_history.py (hunk 8 above)
Comment (craig08) [reply src/langchain_google_memorystore_redis/chat_message_history.py:55]: Redis stores all keys and values in bytes - it encodes every string to bytes with utf-8 so we need to decode it to a native string in Python for input/output

File: src/langchain_google_memorystore_redis/chat_message_history.py (hunk 6 above)
Comment (craig08) [reply src/langchain_google_memorystore_redis/chat_message_history.py]: Thanks for catching this up. The old version of this code manages the client within the class so it closes when it is cleaned up. But we later updated the class to accept user's client, so you're right, we shouldn't close user's client.

File: src/langchain_google_memorystore_redis/chat_message_history.py (hunk 5 above)
Comment (craig08) [reply src/langchain_google_memorystore_redis/chat_message_history.py]: Thanks for bringing this up - I'm not very familiar with the style. Updated

File: src/langchain_google_memorystore_redis/chat_message_history.py (hunk 8 above)
Comment (kurtisvg) [reply src/langchain_google_memorystore_redis/chat_message_history.py:55]: Are we encoding it or is the redis client encoding it (I don't see encode called below)?

Does the redis client explictly call out it encodes in "utf-8"? Or does it use the default for python (which happens to be utf-8 atm)?  https://docs.python.org/3/library/stdtypes.html#bytes.decode

If they are using the default, we should too

**Review by kurtisvg — CHANGES_REQUESTED** [review]

File: src/langchain_google_memorystore_redis/chat_message_history.py (hunk 8 above)
Comment (craig08) [reply src/langchain_google_memorystore_redis/chat_message_history.py]: Great catch. utf-8 is the [default encoding of the Python redis-py package](https://redis-py.readthedocs.io/en/stable/connections.html#generic-client). So we should use the same encoder as the user uses. Updated.

**Review by kurtisvg — APPROVED** [review]

**Pull request merged**